inc/namespace.php

```diff
@@ -957,21 +962,68 @@ function redact_metadata( $metadata ) {
 }
 
 /**
- * Truncate sql query to given maximum size.
+ * Maps array values using a callback recursively.
+ *
+ * @param callable $callback The function to apply to leaf nodes.
+ * @param array $array The array to modify.
+ * @return array
+ */
+function array_map_recursive( $callback, $array ) {
+	$func = function ( $item ) use ( &$func, &$callback ) {
+		return is_array( $item ) ? array_map( $func, $item ) : call_user_func( $callback, $item );
+	};
+
+	return array_map( $func, $array );
+}
+
+/**
+ * Truncate values in the metadata for a trace.
+ *
+ * @param array $trace The trace metadata.
+ * @return array
+ */
+function truncate_trace( $trace ) {
+	$max_size = 63 * 1024; // 63 KB, leaving room for UDP headers etc.
+	$trace_size = mb_strlen( json_encode( $trace ) ); // @codingStandardsIgnoreLine wp_json_encode not available.
+
+	if ( $trace_size <= $max_size ) {
+		return $trace;
+	}
+
+	// Metadata is currently the only target for truncation.
+	if ( ! isset( $trace['metadata'] ) ) {
+		return $trace;
+	}
+
+	// Find the amount of bytes we need to restrict metadata values to.
+	$metadata_size = mb_strlen( json_encode( $trace['metadata'] ) ); // @codingStandardsIgnoreLine wp_json_encode not available.
```

Comment:
If `$trace_size - $metadata_size > $max_size` I think we can return early and not bother with the `array_map_recursive()`? I.e. if the trace is too long due too long subsegments, this function can bail early?

Reply:
If I understand right this is only applied per subsegment already. So `$trace` is really the subsegment. Looks like metadata is sent with every subsegment. Can definitely reduce the number of size checks though. I’ll go over it again.

Reply:
Ok, but doesn't the same apply if other stuff apart from the metadata is > 64kb?

Reply:
Yeah, I think per subsegment that can be a problem, I should rename `$trace` to `$subsegment` perhaps though.

I couldn't really see anything beyond metadata that was safe to redact. `$_POST` for example is one of the main things there's no control over and can have legitimately large values, so it's a step in the right direction.

Reply:
Updated with this check, I think it's about as efficient as it can be. It's a little weird but functional.

Reply:
Ok nice